lib/azure/core/http/http_request.rb

```diff
@@ -0,0 +1,175 @@
+        attr_accessor :headers
+
+        # The body of the request (IO or String)
+        attr_accessor :body
+
+        # Azure client which contains configuration context and http agents
+        # @return [Azure::Client]
+        attr_accessor :client
+
+        # Public: Create the HttpRequest
+        #
+        # @param method   [Symbol] The HTTP method to use (:get, :post, :put, :del, etc...)
+        # @param uri      [URI] The URI of the HTTP endpoint to query
+        # @param options_or_body  [Hash|IO|String] The request options including {:client, :body} or raw body only
+        def initialize(method, uri, options_or_body = {})
+          options ||= unless options_or_body.is_a?(Hash)
+                        {body: options_or_body}
+                      end || options_or_body || {}
+
+          @method = method
+          @uri = if uri.is_a?(String)
+                   URI.parse(uri)
+                 else
+                   uri
+                 end
+
+          @client = options[:client] || Azure
+
+          self.headers = default_headers(options[:current_time] || Time.now.httpdate).merge(options[:headers] || {})
+          self.body = options[:body]
+        end
+
+        # Public: Applies a HttpFilter to the HTTP Pipeline
+        #
+        # filter - Any object that responds to .call(req, _next) and
+        #          returns a HttpResponse eg. HttpFilter, Proc,
+        #          lambda, etc. (optional)
+        #
+        # &block - An inline block may be used instead of a filter
+        #
+        #          example:
+        #
+        #             request.with_filter do |req, _next|
+        #               _next.call
+        #             end
+        #
+        # NOTE:
+        #
+        # The code block provided must call _next or the filter pipeline
+        # will not complete and the HTTP request will never execute
+        #
+        def with_filter(filter=nil, &block)
+          filter = filter || block
+          if filter
+            old_impl = self._method(:call)
+
+            # support 1.8.7 (define_singleton_method doesn't exist until 1.9.1)
+            new_impl = Proc.new do
+              filter.call(self, old_impl)
+            end
+            k = class << self;
+              self;
+            end
+            if k.method_defined? :define_singleton_method
+              self.define_singleton_method(:call, new_impl)
+            else
+              k.send(:define_method, :call, new_impl)
+            end
+          end
+        end
+
+        # Build a default headers Hash
+        def default_headers(current_time)
+          {}.tap do |def_headers|
+            def_headers['User-Agent'] = Azure::Core::Default::USER_AGENT
+            def_headers['x-ms-date'] = current_time
+            def_headers['x-ms-version'] = '2014-02-14'
+            def_headers['DataServiceVersion'] = '1.0;NetFx'
+            def_headers['MaxDataServiceVersion'] = '2.0;NetFx'
```

Comment:
I took this value from the prior code, not sure what implications of updating this would be at the moment. Should we address this as an issue later?